src/compiler/objectConstructors.ts

```diff
@@ -0,0 +1,350 @@
+    //       symbol as the key.
+    mergeId = 0;
+    parent?: Symbol = undefined;
+    members?: SymbolTable = undefined;
+    exports?: SymbolTable = undefined;
+    exportSymbol?: Symbol = undefined;
+
+    // TODO: Can we remove this property or turn it into a flag along with `isReplaceableByMethod`?
+    constEnumOnlyModule: boolean | undefined = undefined;
+    // TODO: Does this property require all symbol flags or a subset? Could we use a different enum and combine it with
+    //       `constEnumOnlyModule` and `isReplaceableByMethod`?
+    isReferenced?: SymbolFlags = undefined;
+    // TODO: This is set by checker and not the binder. It should be moved to `SymbolLinks`.
+    lastAssignmentPos?: number = undefined;
+    links?: SymbolLinks = undefined; // used by TransientSymbol
+
+    // TODO: Review these for polymorphism:
+    declare isReplaceableByMethod?: boolean;
+    declare assignmentDeclarationMembers?: Map<number, Declaration>;
+    declare globalExports?: SymbolTable;
+
+    constructor(flags: SymbolFlags, name: __String) {
+        this.flags = flags;
+        this.escapedName = name;
+    }
+
+    getFlags(): SymbolFlags {
+        return this.flags;
+    }
+
+    get name(): string {
+        return symbolName(this);
+    }
+
+    getEscapedName(): __String {
+        return this.escapedName;
+    }
+
+    getName(): string {
+        return this.name;
+    }
+
+    getDeclarations(): Declaration[] | undefined {
+        return this.declarations;
+    }
+
+    getDocumentationComment(checker: TypeChecker | undefined): SymbolDisplayPart[] {
+        return SymbolObjectInternals.internals.getDocumentationComment(this, checker);
+    }
+
+    getContextualDocumentationComment(context: Node | undefined, checker: TypeChecker | undefined): SymbolDisplayPart[] {
+        return SymbolObjectInternals.internals.getContextualDocumentationComment(this, context, checker);
+    }
+
+    getJsDocTags(checker?: TypeChecker): JSDocTagInfo[] {
+        return SymbolObjectInternals.internals.getJsDocTags(this, checker);
+    }
+
+    getContextualJsDocTags(context: Node | undefined, checker: TypeChecker | undefined): JSDocTagInfo[] {
+        return SymbolObjectInternals.internals.getContextualJsDocTags(this, context, checker);
+    }
+}
+
+/** @internal */
+export class TypeObject implements Type {
+    flags: TypeFlags;
+    checker: TypeChecker;
+    id = 0;
+
+    // TODO: Review for polymorphism
+    declare symbol: Symbol;
+    declare pattern?: DestructuringPattern;
+    declare aliasSymbol?: Symbol;
+    declare aliasTypeArguments?: readonly Type[];
+    declare permissiveInstantiation?: Type;
+    declare restrictiveInstantiation?: Type;
+    declare uniqueLiteralFilledInstantiation?: Type;
+    declare immediateBaseConstraint?: Type;
+    declare widened?: Type;
+
+    constructor(checker: TypeChecker, flags: TypeFlags) {
+        this.flags = flags;
+        this.checker = checker;
+    }
+
+    getFlags(): TypeFlags {

```

Comment:
I do wonder if we need to add some sort of lint rule for ourselves to prevent us from needlessly using the methods in the compiler, but I'm not sure how efficient that would be.

Reply:
I feel like it might be worth it just to prevent accidental usage, since these are defined on the actual `Type` interface too. But it's a little goofy to complain when they are now definitely actually declared in the compiler project? (Some stuff is shimmed out and throws, though...)